prow/plugins/blunderbuss/blunderbuss.go

```diff
@@ -105,72 +98,41 @@ func handle(ghc githubClient, oc ownersClient, log *logrus.Entry, reviewerCount
 	return nil
 }
 
-func chance(val, total int64) float64 {
-	return 100.0 * float64(val) / float64(total)
-}
-
-func getPotentialReviewers(owners ownersClient, author string, files []github.PullRequestChange, leafOnly bool) (weightMap, int64) {
-	potentialReviewers := weightMap{}
-	weightSum := int64(0)
-	var fileOwners sets.String
+func getReviewers(owners ownersClient, author string, files []github.PullRequestChange, minReviewers int) ([]string, error) {
+	authorSet := sets.NewString(author)
+	reviewers := sets.NewString()
+	leafReviewers := sets.NewString()
+	// first build 'reviewers' such that each file has at least one leaf reviewer in the set.
 	for _, file := range files {
-		fileWeight := int64(1)
-		if file.Changes != 0 {
-			fileWeight = int64(file.Changes)
-		}
-		// Judge file size on a log scale-- effectively this
-		// makes three buckets, we shouldn't have many 10k+
-		// line changes.
-		fileWeight = int64(math.Log10(float64(fileWeight))) + 1
-		if leafOnly {
-			fileOwners = owners.LeafReviewers(file.Filename)
-		} else {
-			fileOwners = owners.Reviewers(file.Filename)
+		fileLeaf := owners.LeafReviewers(file.Filename).Difference(authorSet)
+		leafReviewers = leafReviewers.Union(fileLeaf)
+		if fileLeaf.Len() == 0 || reviewers.Intersection(fileLeaf).Len() > 0 {
```

Comment:
Seems like we want this?

possibleReviewers := fileLeaf.Difference(reviewers)
reviewers.Insert(popRandom(fileLeaf))

I don't understand why we continue if fileLeaf has some of the same reviewers (there might be some who are not the same!)

Reply:
If `reviewers.Intersection(fileLeaf).Len() > 0` then the set of reviewers we are building already contains a leaf reviewer for the current file so we don't need to add any more reviewers to satisfy the first requirement (every changed file has at least one of it's leaf reviewers requested).

This is also why it is unnecessary to filter the set of changed files down to the set of OWNERS files.

Reply:
Personally I do not consider minimizing the number of assignees a valuable goal.

If I change foo, bar and other. I would expect to see three reviewers on the PR, one from foo, one from bar and one from other. If A is a reviewer for both foo and bar and we select A for foo, I'd expect us to select a different person for bar...

Otherwise it seems like we're still overlying biasing people who are reviewers all over the place. When what we're trying to optimize for is giving new people experience (not making reviews as efficient as possible).

However feel free to punt.

Thoughts on this logic being worth a unit test?

Reply:
That sounds reasonable to me. I changed the first step to try to choose a unique reviewer for each leaf OWNERS file instead of ensuring that each file had at least one leaf reviewer.
I added some test cases to validate that the behavior has changed as expected.